modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/NodeImpl.java

```diff
@@ -1129,6 +1131,48 @@ public boolean init(final NodeOptions opts) {
         return true;
     }
 
+    /**
+     * If there are externally established peers, then abstains from becoming a leader till next configuration change.
+     *
+     * <p>The idea is that, if a Raft group was forcefully repaired (because it lost majority) using {@link #resetPeers( Configuration)},
+     * the old majority nodes might come back online. If this happens and we do nothing, they might elect a leader from the old majority
+     * that could hijack leadership and cause havoc in the repaired group.
+     *
+     * <p>To prevent this, on a starup, current voting set (aka peers) of the repaired group may be set via
+     * {@link NodeOptions#getExternallyEstablishedPeers()}. If a Raft node sees at startup that, according to its log it's a member of
+     * the voting set, but not a member of externally established peers, it realizes that it should abstain from becoming a leader to avoid
+     * a hijack.
+     *
+     * <p>If the group is reconfigured according to the Raft protocol (that is, non-forcefully), a Raft node might see that it's not
+     * a member of the externally established peers anymore and start abstaining from becoming a leader as well, but that's not a problem
+     * if the externally established peers are ONLY updated after the corresponding configuration change is committed as in this case
+     * a majority of the voting set already stores the new configuration in its log, so only a minority of nodes can be abstaining,
+     * which will not create troubles with group availability due to abstaining.
+     */
+    private void initLeadershipAbstaining() {
+        if (options.getExternallyEstablishedPeers() == null) {
+            return;
+        }
+
+        boolean thisNodeInVotingSet = this.conf.listPeers().stream().anyMatch(this.serverId::equals);
+        boolean abstainFromLeadership = thisNodeInVotingSet
+                && !options.getExternallyEstablishedPeers().contains(this.serverId.getConsistentId());
+
+        if (abstainFromLeadership) {
+            LOG.info("Will abstain from becoming a leader until a configuration update [nodeId={}, externallyEnforcedPeers={}].",
+                    this.nodeId, options.getExternallyEstablishedPeers()
+            );
+
+            List<PeerId> newPeers = options.getExternallyEstablishedPeers().stream()
+                    .map(PeerId::new)
+                    .collect(toList());
+            Configuration newConf = new Configuration(newPeers);
+            LOG.info("Node {} set peers from {} to {} to abstain from becoming a leader.", getNodeId(), this.conf.getConf(), newPeers);
+            this.conf.setConf(newConf);
```

Comment:
Should we also update ConfigurationManager?

Reply:
I think we shouldn't. We update the configuration in memory just to make it impossible for the node to become a leader, and it seems to be enough.

Reply:
Probably you are right.